sdk/tools/spec2def/spec2def.c

```diff
@@ -711,52 +744,151 @@ OutputLine_def(FILE *fileDest, EXPORT *pexp)
+    vfprintf(stderr, format, argptr);
+    fprintf(stderr, "\n");
+
+    /* Output the line with the error */
+    fprintf(stderr, "> %.*s\n", len, pcLine);
+
+    if (errorlen == 0)
+    {
+        errorlen = TokenLength(pc);
+    }
+
+    for (i = 0; i < errorpos + 2; i++)
+    {
+        fprintf(stderr, " ");
+    }
+    for (i = 0; i < errorlen; i++)
+    {
+        fprintf(stderr, "~");
+    }
+    fprintf(stderr, "\n");
+    exit(-1);
+}
+
+void
+Fatal(
+    const char* filename,
+    unsigned nLine,
+    const char *pcLine,
+    const char *pc,
+    unsigned errorlen,
+    const char *format,
+    ...)
+{
+    va_list argptr;
+
+    va_start(argptr, format);
+    Fatalv(filename, nLine, pcLine, pc, errorlen, format, argptr);
+    va_end(argptr);
+}
+
+EXPORT *
+ParseFile(char* pcStart, FILE *fileDest, unsigned *cExports)
+{
+    EXPORT *pexports;
+    const char *pc, *pcLine;
+    int cLines, nLine;
     EXPORT exp;
-    int included, version_included;
-    char namebuffer[16];
+    int included;
     unsigned int i;
 
+    *cExports = 0;
+
     //fprintf(stderr, "info: line %d, pcStart:'%.30s'\n", nLine, pcStart);
 
+    /* Count the lines */
+    for (cLines = 1, pcLine = pcStart; *pcLine; pcLine = NextLine(pcLine), cLines++)
+    {
+        /* Nothing */
+    }
+
+    /* Allocate an array of EXPORT structures */
+    pexports = (EXPORT*)malloc(cLines * sizeof(EXPORT));
+    if (pexports == NULL)
+    {
+        fprintf(stderr, "ERROR: %s: failed to allocate EXPORT array of %u elements\n", pszSourceFileName, cLines);
+        return 0;
+    }
+
     /* Loop all lines */
     nLine = 1;
     exp.nNumber = 0;
     for (pcLine = pcStart; *pcLine; pcLine = NextLine(pcLine), nLine++)
     {
         pc = pcLine;
 
+        exp.strName.buf = NULL;
+        exp.strName.len = 0;
+        exp.strTarget.buf = NULL;
+        exp.strTarget.len = 0;
         exp.nArgCount = 0;
         exp.uFlags = 0;
         exp.nNumber++;
+        exp.nStartVersion = 0;
+        exp.nEndVersion = 0xFFFFFFFF;
+        exp.bVersionIncluded = 1;
 
         /* Skip white spaces */
         while (*pc == ' ' || *pc == '\t') pc++;
 
-        /* Skip empty lines, stop at EOF */
+        /* Skip empty lines */
         if (*pc == ';' || *pc <= '#') continue;
-        if (*pc == 0) return 0;
+
+        /* On EOF we are done */
+        if (*pc == 0)
+        {
+            return pexports;
+        }
 
         /* Now we should get either an ordinal or @ */
         if (*pc == '@')
+        {
             exp.nOrdinal = -1;
-        else
+        }
+        else if ((*pc >= '0') && (*pc <= '9'))
         {
             exp.nOrdinal = atol(pc);
             /* The import lib should contain the ordinal only if -ordinal was specified */
             if (!gbImportLib)
                 exp.uFlags |= FL_ORDINAL;
         }
+        else
+        {
+            Fatal(pszSourceFileName, nLine, pcLine, pc, 0, "Expected '@' or ordinal");
```

Comment:
Arguably `errorlen` should be 1 here since we looked at only one char to determine that this is an error.

Reply:
But it marks the entire token, i.e. if this was ```  -4 stdcall foo()``` it would mark the ```-4```